### PR TITLE
HDFS-9913. DistCp to add -useTrash to move deleted files to Trash

### PR DESCRIPTION
**Pull request by asagjj**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 75 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 5 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1327 | trunk passed |
| +1 | compile | 28 | trunk passed |
| +1 | checkstyle | 21 | trunk passed |
| +1 | mvnsite | 30 | trunk passed |
| +1 | shadedclient | 787 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | trunk passed |
| 0 | spotbugs | 45 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 43 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 25 | the patch passed |
| +1 | compile | 23 | the patch passed |
| +1 | javac | 23 | the patch passed |
| -0 | checkstyle | 16 | hadoop-tools/hadoop-distcp: The patch generated 6 new + 136 unchanged - 0 fixed = 142 total (was 136) |
| +1 | mvnsite | 26 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 828 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 19 | the patch passed |
| +1 | findbugs | 51 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 834 | hadoop-distcp in the patch failed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 4297 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.tools.contract.TestLocalContractDistCp |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-974/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/974 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux cf29b11bc491 4.4.0-141-generic #167~14.04.1-Ubuntu SMP Mon Dec 10 13:20:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 0766186 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-distcp.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/1/artifact/out/patch-unit-hadoop-tools_hadoop-distcp.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/1/testReport/ |
| Max. process+thread count | 305 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 75 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 5 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1238 | trunk passed |
| +1 | compile | 29 | trunk passed |
| +1 | checkstyle | 21 | trunk passed |
| +1 | mvnsite | 31 | trunk passed |
| +1 | shadedclient | 774 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | trunk passed |
| 0 | spotbugs | 46 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 42 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 27 | the patch passed |
| +1 | compile | 24 | the patch passed |
| +1 | javac | 24 | the patch passed |
| -0 | checkstyle | 16 | hadoop-tools/hadoop-distcp: The patch generated 1 new + 136 unchanged - 0 fixed = 137 total (was 136) |
| +1 | mvnsite | 26 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 811 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 20 | the patch passed |
| +1 | findbugs | 49 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 892 | hadoop-distcp in the patch passed. |
| +1 | asflicense | 34 | The patch does not generate ASF License warnings. |
| | | 4232 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-974/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/974 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 46a9b8f0b152 4.15.0-48-generic #51-Ubuntu SMP Wed Apr 3 08:28:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 304a47e |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-distcp.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/2/testReport/ |
| Max. process+thread count | 342 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-974/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by asagjj** [comment]
@steveloughran @jojochuang  Could you please help to review this commit?

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyCommitter.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 package org.apache.hadoop.tools.mapred;
 
+import org.apache.hadoop.fs.Trash;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyCommitter.java]: nit: should go with the rest of the org.apache imports

File: hadoop-tools/hadoop-distcp/src/site/markdown/DistCp.md.vm [hunk 2]
```diff
@@ -349,7 +349,7 @@ Command Line Options
 | `-filters` | The path to a file containing a list of pattern strings, one string per line, such that paths matching the pattern will be excluded from the copy. | Support regular expressions specified by java.util.regex.Pattern. |
 | `-filelimit <n>` | Limit the total number of files to be <= n | **Deprecated!** Ignored in the new DistCp. |
 | `-sizelimit <n>` | Limit the total size to be <= n bytes | **Deprecated!** Ignored in the new DistCp. |
-| `-delete` | Delete the files existing in the dst but not in src | The deletion is done by FS Shell. So the trash will be used, if it is enable. Delete is applicable only with update or overwrite options. |
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/site/markdown/DistCp.md.vm]: Typo in "dst"

File: hadoop-tools/hadoop-distcp/src/site/markdown/DistCp.md.vm [hunk 3]
```diff
@@ -349,7 +349,7 @@ Command Line Options
...
+| `-delete [-useTrash]` | Delete the file`[-useTrash]` is enabled,the files will be moved to trash. Notice that `[-useTrash]` option on some object store does a copy and delete ops and costs O(Num). Delete is applicable only with update or overwrite options. |
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/site/markdown/DistCp.md.vm]: 1.  add space between , and "the"
2. for cost of object stores, just say "can be slow"

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpOptions.java [hunk 4]
```diff
@@ -225,6 +225,45 @@ public void testSetDeleteMissing() {
     }
   }
 
+  @Test
+  public void testDeleteMissingUseTrash() {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpOptions.java]: All assert true/false calls *MUST* have a string message to include on a failure. 

Imagine you've got a report of a test run failure on jenkins, you want to understand what has failed without trying to work out exactly which line does what on that specific branch of the code

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpOptions.java [hunk 5]
```diff
@@ -225,6 +225,45 @@ public void testSetDeleteMissing() {
...
+        new Path("hdfs://localhost:8020/target/"))
+        .withOverwrite(true)
+        .withDeleteMissing(true)
+        .withDeleteUseTrash(true)
+        .build();
+    Assert.assertTrue(options.shouldDeleteUseTrash());
+    Assert.assertTrue(options.shouldOverwrite());
+    Assert.assertTrue(options.shouldDeleteMissing());
+
+    try {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestDistCpOptions.java]: Use `LambdaTestUtils.intercept()` here; there are plenty other examples in the code

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestOptionsParser.java [hunk 6]
```diff
@@ -802,4 +802,21 @@ public void testExclusionsOption() {
         "hdfs://localhost:8020/target/"});
     Assert.assertEquals(options.getFiltersFile(), "/tmp/filters.txt");
   }
+
+  @Test
+  public void testParseDeleteSkipTrash() {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/TestOptionsParser.java]: again, asserts need messages. sorry

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/contract/AbstractContractDistCpTest.java [hunk 7]
```diff
@@ -322,6 +335,75 @@ private Job distCpUpdate(final Path srcDir, final Path destDir)
             .withOverwrite(false)));
   }
 
+  /**
+   * Do a distcp -update -delete -useTrash.
+   * @param destDir output directory used by the initial distcp
+   * @return the distcp job
+   */
+
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/contract/AbstractContractDistCpTest.java]: nit. cut this empty line

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/contract/AbstractContractDistCpTest.java [hunk 8]
```diff
@@ -322,6 +335,75 @@ private Job distCpUpdate(final Path srcDir, final Path destDir)
...
+        "Paths for test are wrong",
+        inputFile1, inputFile2, inputFile3, inputFile4, inputFile5);
+
+    modifySourceDirectories();
+    ContractTestUtils.assertPathsDoNotExist(localFS, "deleted right now",
+        inputFile1, inputFile3, inputSubDir4);
+
+    Path trashRootDir = remoteFS.getTrashRoot(null);
+    if (remoteFS.exists(trashRootDir)) {
+      remoteFS.delete(trashRootDir, true);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/contract/AbstractContractDistCpTest.java]: Use `ContractTestUtils.assertDeleted()`

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestCopyCommitter.java [hunk 9]
```diff
@@ -18,6 +18,7 @@
 
 package org.apache.hadoop.tools.mapred;
 
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestCopyCommitter.java]: nit: move to rest of org.apache imports

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestCopyCommitter.java [hunk 10]
```diff
@@ -204,6 +206,60 @@ public void testDeleteMissing() throws IOException {
...
+    String sourceBase;
+    String targetBase;
+    FileSystem fs = null;
+    try {
+      OutputCommitter committer = new CopyCommitter(null, taskAttemptContext);
+      fs = FileSystem.get(conf);
+      sourceBase = TestDistCpUtils.createTestSetup(fs);
+      targetBase = TestDistCpUtils.createTestSetup(fs);
+      String targetBaseAdd = TestDistCpUtils.createTestSetup(fs);
+      fs.rename(new Path(targetBaseAdd), new Path(targetBase));
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestCopyCommitter.java]: `ContractTestUtils.assertRenameOutcome()`

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpOptionSwitch.java [hunk 11]
```diff
@@ -76,6 +76,14 @@
       new Option("delete", false, "Delete from target, " +
           "files missing in source. Delete is applicable only with update or overwrite options")),
 
+  /**
+   * When -delete option on, files in target that are missing from source
+   * will be delete by default. This allows the files to be
+   * moved to the trash
+   */
+  DELETE_USETRASH(DistCpConstants.CONF_LABEL_DELETE_MISSING_USETRASH,
+      new Option("useTrash", false, "Delete from target to the trash")),
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpOptionSwitch.java]: Or "Move deleted files into the user's trash directory in the destination filesystem"

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpOptions.java [hunk 12]
```diff
@@ -284,6 +287,14 @@ public boolean shouldUseSnapshotDiff() {
     return shouldUseDiff() || shouldUseRdiff();
   }
 
+  /**
+   * Should the file missing in source should be deleted use trash?
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpOptions.java]: hard to parse

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/OptionsParser.java [hunk 13]
```diff
@@ -153,6 +155,9 @@ public static DistCpOptions parse(String[] args)
               command,
               DistCpOptionSwitch.TRACK_MISSING.getSwitch())));
     }
+    if(command.hasOption(DistCpOptionSwitch.DELETE_USETRASH.getSwitch())) {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/OptionsParser.java]: nit: space between if and (

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyCommitter.java [hunk 14]
```diff
@@ -515,6 +518,17 @@ private void deleteMissing(Configuration conf) throws IOException {
         formatDuration(deletionEnd - listingStart));
   }
 
+  private boolean deletePath(FileSystem targetFS, Path targetEntry,
+                             Configuration conf) throws IOException {
+    if(conf.getBoolean(DistCpConstants.CONF_LABEL_DELETE_MISSING_USETRASH,
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/CopyCommitter.java]: nit: add  whitespace after `if`

**Comment by steveloughran** [comment]
Production code looks good other than a few minor issues; commented more on the tests.

Have you run any of the object store tests runs (abfs, s3a, etc.)? Not a blocker but I'd prefer this, as otherwise I'll end up doing it

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpOptions.java (hunk 12 above)
Comment (asagjj) [reply hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DistCpOptions.java]: remove java doc

**Comment by asagjj** [comment]
Thanks @steveloughran  for patient review!!  I had ran and passed all mentioned object stores(abfs, azure, s3a,adl, aliyun) unit tests. I will update commits according to you suggestions soon.

**Comment by steveloughran** [comment]
OK, +1 from me

**Comment by steveloughran** [comment]
committed; extracted your email address from the .patch file so that you are credited as the author of the patch.

**Comment by steveloughran** [comment]
Update: JIRA playing up so can't close the issue. 

But I'm also getting a failure on my test run
```
[INFO] Running org.apache.hadoop.fs.s3a.scale.ITestS3AConcurrentOps
[ERROR] Tests run: 9, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 223.958 s <<< FAILURE! - in org.apache.hadoop.fs.contract.s3a.ITestS3AContractDistCp
[ERROR] testUpdateUseTrashDeepDirectoryStructureToRemote(org.apache.hadoop.fs.contract.s3a.ITestS3AContractDistCp)  Time elapsed: 20.398 s  <<< ERROR!
java.io.FileNotFoundException: about to be deleted file: not found s3a://hwdev-steve-ireland-new/user/stevel/.Trash in s3a://hwdev-steve-ireland-new/user/stevel
	at org.apache.hadoop.fs.contract.ContractTestUtils.verifyPathExists(ContractTestUtils.java:942)
	at org.apache.hadoop.fs.contract.ContractTestUtils.assertPathExists(ContractTestUtils.java:920)
	at org.apache.hadoop.fs.contract.ContractTestUtils.assertDeleted(ContractTestUtils.java:704)
	at org.apache.hadoop.fs.contract.ContractTestUtils.assertDeleted(ContractTestUtils.java:686)
	at org.apache.hadoop.tools.contract.AbstractContractDistCpTest.distCpUpdateUseTrashDeepDirectoryStructure(AbstractContractDistCpTest.java:358)
	at org.apache.hadoop.tools.contract.AbstractContractDistCpTest.testUpdateUseTrashDeepDirectoryStructureToRemote(AbstractContractDistCpTest.java:233)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.FileNotFoundException: No such file or directory: s3a://hwdev-steve-ireland-new/user/stevel/.Trash
	at org.apache.hadoop.fs.s3a.S3AFileSystem.s3GetFileStatus(S3AFileSystem.java:2837)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.innerGetFileStatus(S3AFileSystem.java:2726)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.getFileStatus(S3AFileSystem.java:2620)
	at org.apache.hadoop.fs.contract.ContractTestUtils.verifyPathExists(ContractTestUtils.java:936)
	... 21 more
```

I'll look at this a bit more and decide whether we should fix this with rollback vs another iteration of the PR

**Comment by steveloughran** [comment]
OK, cause is that the test setup calls `assertDeleted()` to clean up the trash dir, and the CTU check there actually fails if the file doesn't exist before the delete. Which is actually something we should question. 

For now: reverted this PR so everything will be working before the US wakes up. I know it's a quick fix but I'd be depending on others to review, etc, etc and the failing test will be around for 24h minimum.

@asagjj : Can you submit an updated PR with the delete setup something like:

```java
    Path trashRootDir = remoteFS.getTrashRoot(null);
    remoteFS.delete(trashRootDir, true);
    ContractTestUtils.assertPathDoesNotExist(remoteFS, "trash",  trashRootDir )
```

We don't need to worry about the before state or the return value of delete, just that the dir isn't there. 
Thanks

**Comment by asagjj** [comment]
@steveloughran  Sorry about that, I didn't reproduce this failure in my enviroment. And since this pr is closed ,I open a new one to update the commit [https://github.com/apache/hadoop/pull/1111](url)

**Comment by steveloughran** [comment]
thx, I'll look at it tomorrow. I'll run the tests first just to make sure all is good.
